### PR TITLE
Implement comparisons in sqlalchemy.

### PR DESCRIPTION
**Pull request by dhatch**
PR checklist:

- [x] Added changelog entry.

### CODE REVIEW

File: languages/python/django-oso/django_oso/partial.py [hunk 1]
```diff
@@ -91,17 +91,23 @@ def and_expr(expr: Expression, model: Model, **kwargs):
...
     left_path = dot_path(left)
     if left_path:
         return COMPARISONS[expr.operator](q, "__".join(path + left_path), right)
     else:
-        if isinstance(right, Model):
+        if isinstance(right, model):
             right = right.pk
         else:
+            # QUESTION: Should this just return an empty set? Similar to how invalid isa
+            # returns empty set.
```
Comment (gj) [inline languages/python/django-oso/django_oso/partial.py:105]: Feels different to me -- the semantics of returning an empty set would be "this is a valid query that returns no results," which I don't think is an accurate representation of what's happening.

File: languages/python/django-oso/django_oso/partial.py [hunk 2]
```diff
@@ -91,17 +91,23 @@ def and_expr(expr: Expression, model: Model, **kwargs):
...
             raise UnsupportedError(f"Unsupported comparison: {expr}")
+
+        if not expr.operator in ('Eq', 'Unify'):
+            raise UnsupportedError(f"Comparing model only supported with Eq or Unify not: {expr}")
```
Comment (gj) [inline languages/python/django-oso/django_oso/partial.py:109]: Why is this necessary?

File: languages/python/django-oso/django_oso/partial.py [hunk 3]
```diff
@@ -91,17 +91,23 @@ def and_expr(expr: Expression, model: Model, **kwargs):
...
-def compare_expr(expr: Expression, _model: Model, path=(), **kwargs):
+def compare_expr(expr: Expression, model: Model, path=(), **kwargs):
     q = Q()
     (left, right) = expr.args
     left_path = dot_path(left)
     if left_path:
         return COMPARISONS[expr.operator](q, "__".join(path + left_path), right)
     else:
-        if isinstance(right, Model):
+        if isinstance(right, model):
```
Comment (gj) [inline languages/python/django-oso/django_oso/partial.py:101]: I don't think this is correct. All the previous code is doing is translating any Django model into its primary key. We pass the authorized model around as `model`, so this change would cause comparisons on subfields that don't inherit from `model` to fail.

Comment (dhatch) [reply languages/python/django-oso/django_oso/partial.py:101]: I should add a test for this.  This branch (from my understanding) only runs for something like `this = post` where `post` is a variable with the `Post` type.  I changed the check because conceivably you could write `this = user`.  This doesn't do any type checking so if the types don't match, we still extract a PK, and then we'd be trying to find a Post that has the PK of the user.

File: languages/python/django-oso/django_oso/partial.py (hunk 1 above)
Comment (dhatch) [reply languages/python/django-oso/django_oso/partial.py:105]: Not sure... see my example above.  Unifying something with a type that doesn't match would end up in this branch, which is very similar to checking isa with a type that doesn't match.

File: languages/python/django-oso/django_oso/partial.py (hunk 2 above)
Comment (dhatch) [reply languages/python/django-oso/django_oso/partial.py:109]: Because it doesn't make sense otherwise.  If you write `this < post`, you would probably expect to use the natural ordering of post, which isn't something we know about, not to get all posts that have a primary key less than that of the right hand side.

File: languages/python/django-oso/django_oso/partial.py (hunk 3 above)
Comment (gj) [reply languages/python/django-oso/django_oso/partial.py:101]: I see where you're coming from -- if `left_path` is falsy then there's no dot operation and the LHS must be `_this`, so the only reasonably comparable thing is `model`.

Comment (dhatch) [reply languages/python/django-oso/django_oso/partial.py:101]: Exactly

**Comment by dhatch** [comment]
Enumerated tests to add for this here: https://www.notion.so/osohq/TEST-Test-for-model-comparison-this-in-SQLAlchemy-others-68aad30b405e484aa6d058407a52289d

File: languages/python/django-oso/django_oso/partial.py (hunk 1 above)
Comment (gj) [reply languages/python/django-oso/django_oso/partial.py:105]: In light of the above realization, I agree that this should probably return `Q(pk__in=[])`, which is Django's version of the "always false" SQL clause. (See usage in `isa_expr()`, above.)

File: languages/python/django-oso/django_oso/partial.py (hunk 2 above)
Comment (gj) [reply languages/python/django-oso/django_oso/partial.py:109]: Cool, can we use the error message from above (`Unsupported comparison: {expr}`) once we replace that with the `Q(pk__in=[])` thing?

File: languages/python/sqlalchemy-oso/sqlalchemy_oso/partial.py [hunk 4]
```diff
@@ -61,6 +61,10 @@ def translate_compare(expression: Expression, session: Session, model):
     else:
         assert left == Variable("_this")
         assert inspect(right)
+        assert isinstance(right, model)
+
+        if not expression.operator in ('Eq', 'Unify'):
+            raise UnsupportedError(f"Comparing model only supported with Eq or Unify not: {expression.operator}")
```
Comment (gj) [inline languages/python/sqlalchemy-oso/sqlalchemy_oso/partial.py:67]:
```suggestion
            raise UnsupportedError(f"Unsupported comparison: {expression}")
```

File: languages/python/django-oso/django_oso/partial.py (hunk 2 above)
Comment (gj) [inline languages/python/django-oso/django_oso/partial.py:109]:
```suggestion
            raise UnsupportedError(f"Unsupported comparison: {expr}")
```

File: languages/python/sqlalchemy-oso/sqlalchemy_oso/partial.py [hunk 5]
```diff
@@ -61,6 +61,12 @@ def translate_compare(expression: Expression, session: Session, model):
     else:
         assert left == Variable("_this")
         assert inspect(right)
+        assert isinstance(right, model)
+
+        if not expression.operator in ("Eq", "Unify"):
+            raise UnsupportedError(
+                f"Comparing model only supported with Eq or Unify not: {expression.operator}"
+            )
```
Comment (gj) [inline languages/python/sqlalchemy-oso/sqlalchemy_oso/partial.py:69]:
```suggestion
        if expression.operator not in ("Eq", "Unify"):
            raise UnsupportedError(f"Unsupported comparison: {expression}")
```

File: languages/python/django-oso/django_oso/partial.py [hunk 6]
```diff
@@ -91,17 +91,25 @@ def and_expr(expr: Expression, model: Model, **kwargs):
...
             right = right.pk
         else:
+            # QUESTION: Should this just return an empty set? Similar to how invalid isa
+            # returns empty set.
             raise UnsupportedError(f"Unsupported comparison: {expr}")
+
+        if not expr.operator in ("Eq", "Unify"):
+            raise UnsupportedError(
+                f"Comparing model only supported with Eq or Unify not: {expr}"
+            )
```
Comment (gj) [inline languages/python/django-oso/django_oso/partial.py:111]:
```suggestion
        if expr.operator not in ("Eq", "Unify"):
            raise UnsupportedError(f"Unsupported comparison: {expr}")
```

File: languages/python/django-oso/django_oso/partial.py [hunk 7]
```diff
@@ -91,17 +91,25 @@ def and_expr(expr: Expression, model: Model, **kwargs):
...
     if left_path:
         return COMPARISONS[expr.operator](q, "__".join(path + left_path), right)
     else:
-        if isinstance(right, Model):
+        if isinstance(right, model):
             right = right.pk
         else:
+            # QUESTION: Should this just return an empty set? Similar to how invalid isa
+            # returns empty set.
             raise UnsupportedError(f"Unsupported comparison: {expr}")
```
Comment (gj) [inline languages/python/django-oso/django_oso/partial.py:106]: we should probably create a "constant" like `always_false = Q(pk__in=[])` at the top of the file and then use it in both places
```suggestion
			return Q(pk__in=[])
```

File: languages/python/django-oso/django_oso/partial.py (hunk 2 above)
Comment (dhatch) [reply languages/python/django-oso/django_oso/partial.py]: I used a different message because it is more specific and tells you that a Unification or Equality comparison would be valid, not just that the comparison is unsupported.

Comment (dhatch) [reply languages/python/django-oso/django_oso/partial.py]: Revised to match other error more closely.

File: languages/python/django-oso/django_oso/partial.py (hunk 7 above)
Comment (dhatch) [reply languages/python/django-oso/django_oso/partial.py]: I saw you guys added this constant in another PR, we can replace it then.

**Review by gj — APPROVED** [review]
~Align SQLAlchemy error w/ Django error (https://github.com/osohq/oso/pull/568/files#r534404138)~ and then :shipit:
